Cannot update documents with Strings as IDs

### Expected Behavior

Either of:
- saveAll updates all the documents even if they exists in the database
- updateAll updates all the documents and inserts them if they do not exist
- @MongoUpdateOptions(upsert = true) on saveAll method makes it update even if they exist
- @MongoUpdateOptions(upsert = true) on updateAll makes it create documents if they do not exist

### Actual Behaviour

- saveAll -> duplicate key error
- updateAll -> "not a valid ObjectId representation" error see stacktrace below
- @MongoUpdateOptions(upsert = true) on saveAll -> duplicate key error
- @MongoUpdateOptions(upsert = true) on updateAll -> "not a valid ObjectId...."

```
io.micronaut.configuration.kafka.exceptions.KafkaListenerException: invalid hexadecimal representation of an ObjectId: [c88fb9a5977ca1378c44c5b45000fb2b]
	at io.micronaut.configuration.kafka.processor.KafkaConsumerProcessor.handleException(KafkaConsumerProcessor.java:693)
	at io.micronaut.configuration.kafka.processor.KafkaConsumerProcessor.createConsumerThreadPollLoop(KafkaConsumerProcessor.java:488)
	at io.micronaut.configuration.kafka.processor.KafkaConsumerProcessor.lambda$submitConsumerThread$7(KafkaConsumerProcessor.java:421)
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:79)
	at io.micrometer.core.instrument.Timer.lambda$wrap$0(Timer.java:160)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: invalid hexadecimal representation of an ObjectId: [c88fb9a5977ca1378c44c5b45000fb2b]
	at org.bson.types.ObjectId.parseHexString(ObjectId.java:419)
	at org.bson.types.ObjectId.<init>(ObjectId.java:205)
	at io.micronaut.data.mongodb.operations.MongoUtils.idValue(MongoUtils.java:132)
	at io.micronaut.data.mongodb.operations.MongoUtils.entityIdValue(MongoUtils.java:110)
	at io.micronaut.data.mongodb.operations.MongoUtils.filterByEntityId(MongoUtils.java:152)
	at io.micronaut.data.mongodb.operations.MongoUtils.filterByIdAndVersion(MongoUtils.java:167)
	at io.micronaut.data.mongodb.operations.DefaultMongoRepositoryOperations$5.lambda$collectAutoPopulatedPreviousValues$1(DefaultMongoRepositoryOperations.java:855)
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:180)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at io.micronaut.data.mongodb.operations.DefaultMongoRepositoryOperations$5.collectAutoPopulatedPreviousValues(DefaultMongoRepositoryOperations.java:855)
	at io.micronaut.data.runtime.operations.internal.BaseOperations.update(BaseOperations.java:113)
	at io.micronaut.data.runtime.operations.internal.SyncEntitiesOperations.update(SyncEntitiesOperations.java:33)
	at io.micronaut.data.mongodb.operations.DefaultMongoRepositoryOperations.lambda$updateAll$14(DefaultMongoRepositoryOperations.java:549)
	at io.micronaut.data.mongodb.operations.DefaultMongoRepositoryOperations.withClientSession(DefaultMongoRepositoryOperations.java:761)
	at io.micronaut.data.mongodb.operations.DefaultMongoRepositoryOperations.updateAll(DefaultMongoRepositoryOperations.java:539)
	at io.micronaut.data.runtime.intercept.DefaultUpdateAllEntitiesInterceptor.intercept(DefaultUpdateAllEntitiesInterceptor.java:48)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:135)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:98)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:138)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at de.rhenus.fl.dpx.shipment.adapter.mongo.shipment.ShipmentRepository$Intercepted.updateAll(Unknown Source)
```



### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

https://github.com/ckosmowski/micronaut-examples/tree/issues/mongodb/stringids

### Version

3.4.4

--- a/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/AbstractMongoRepositoryOperations.java
+++ b/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/AbstractMongoRepositoryOperations.java
@@ -16,6 +16,10 @@
 package io.micronaut.data.mongodb.operations;
 
 import com.mongodb.client.model.Collation;
+import com.mongodb.client.model.DeleteOptions;
+import com.mongodb.client.model.InsertManyOptions;
+import com.mongodb.client.model.InsertOneOptions;
+import com.mongodb.client.model.ReplaceOptions;
 import io.micronaut.aop.MethodInvocationContext;
 import io.micronaut.context.BeanContext;
 import io.micronaut.core.annotation.AnnotationMetadata;
@@ -25,6 +29,7 @@ import io.micronaut.core.convert.ConversionService;
 import io.micronaut.core.type.Argument;
 import io.micronaut.core.util.StringUtils;
 import io.micronaut.data.annotation.Repository;
+import io.micronaut.data.intercept.RepositoryMethodKey;
 import io.micronaut.data.intercept.annotation.DataMethod;
 import io.micronaut.data.model.Pageable;
 import io.micronaut.data.model.runtime.AttributeConverterRegistry;
@@ -37,6 +42,7 @@ import io.micronaut.data.model.runtime.StoredQuery;
 import io.micronaut.data.mongodb.annotation.MongoRepository;
 import io.micronaut.data.mongodb.operations.options.MongoAggregationOptions;
 import io.micronaut.data.mongodb.operations.options.MongoFindOptions;
+import io.micronaut.data.mongodb.operations.options.MongoOptionsUtils;
 import io.micronaut.data.operations.HintsCapableRepository;
 import io.micronaut.data.repository.GenericRepository;
 import io.micronaut.data.runtime.config.DataSettings;
@@ -52,6 +58,7 @@ import io.micronaut.inject.BeanDefinition;
 import io.micronaut.inject.ExecutableMethod;
 import io.micronaut.inject.qualifiers.Qualifiers;
 import org.bson.BsonDocument;
+import org.bson.BsonDocumentWrapper;
 import org.bson.BsonNull;
 import org.bson.BsonValue;
 import org.bson.codecs.configuration.CodecRegistry;
@@ -64,6 +71,8 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.stream.Collectors;
 
 /**
@@ -95,6 +104,8 @@ abstract class AbstractMongoRepositoryOperations<Dtb, Cnt, PS> extends AbstractR
             return conversionService;
         }
     };
+
+    private final ConcurrentMap<RepositoryMethodKey, Object> options = new ConcurrentHashMap<>(50);
 
     /**
      * Default constructor.
@@ -128,6 +139,22 @@ abstract class AbstractMongoRepositoryOperations<Dtb, Cnt, PS> extends AbstractR
             }
         }
         this.repoDatabaseConfig = Collections.unmodifiableMap(repoDatabaseConfig);
+    }
+
+    protected final ReplaceOptions getReplaceOptions(AnnotationMetadata annotationMetadata) {
+        return MongoOptionsUtils.buildReplaceOptions(annotationMetadata).orElseGet(ReplaceOptions::new);
+    }
+
+    protected final InsertOneOptions getInsertOneOptions(AnnotationMetadata annotationMetadata) {
+        return MongoOptionsUtils.buildInsertOneOptions(annotationMetadata).orElseGet(InsertOneOptions::new);
+    }
+
+    protected final InsertManyOptions getInsertManyOptions(AnnotationMetadata annotationMetadata) {
+        return MongoOptionsUtils.buildInsertManyOptions(annotationMetadata).orElseGet(InsertManyOptions::new);
+    }
+
+    protected final DeleteOptions getDeleteOptions(AnnotationMetadata annotationMetadata) {
+        return MongoOptionsUtils.buildDeleteOptions(annotationMetadata, true).orElseGet(DeleteOptions::new);
     }
 
     protected abstract Dtb getDatabase(RuntimePersistentEntity<?> persistentEntity, Class<?> repository);
@@ -274,14 +301,24 @@ abstract class AbstractMongoRepositoryOperations<Dtb, Cnt, PS> extends AbstractR
         return ConversionContext.DEFAULT;
     }
 
+    protected final <T> Bson createFilterIdAndVersion(RuntimePersistentEntity<T> persistentEntity, T entity, CodecRegistry codecRegistry) {
+        BsonDocument bsonDocument = BsonDocumentWrapper.asBsonDocument(entity, codecRegistry);
+        BsonDocument filter = new BsonDocument();
+        filter.put(MongoUtils.ID, bsonDocument.get(MongoUtils.ID));
+        RuntimePersistentProperty<T> version = persistentEntity.getVersion();
+        if (version != null) {
+            filter.put(version.getPersistedName(), bsonDocument.get(version.getPersistedName()));
+        }
+        return filter;
+    }
+
     protected void logFind(MongoFind find) {
-        StringBuilder sb = new StringBuilder();
+        StringBuilder sb = new StringBuilder("Executing Mongo 'find'");
         MongoFindOptions options = find.getOptions();
         if (options != null) {
+            sb.append(" with");
             Bson filter = options.getFilter();
-            if (filter != null) {
-                sb.append(" filter: ").append(filter.toBsonDocument().toJson());
-            }
+            sb.append(" filter: ").append(filter == null ? "{}" : filter.toBsonDocument().toJson());
             Bson sort = options.getSort();
             if (sort != null) {
                 sb.append(" sort: ").append(sort.toBsonDocument().toJson());
@@ -295,28 +332,21 @@ abstract class AbstractMongoRepositoryOperations<Dtb, Cnt, PS> extends AbstractR
                 sb.append(" collation: ").append(collation);
             }
         }
-        if (sb.length() == 0) {
-            QUERY_LOG.debug("Executing Mongo 'find'");
-        } else {
-            QUERY_LOG.debug("Executing Mongo 'find' with" + sb);
-        }
+        QUERY_LOG.debug(sb.toString());
     }
 
     protected void logAggregate(MongoAggregation aggregation) {
         MongoAggregationOptions options = aggregation.getOptions();
-        StringBuilder sb = new StringBuilder();
+        StringBuilder sb = new StringBuilder("Executing Mongo 'aggregate'");
         if (options != null) {
+            sb.append(" with");
             sb.append(" pipeline: ").append(aggregation.getPipeline().stream().map(e -> e.toBsonDocument().toJson()).collect(Collectors.toList()));
             Collation collation = options.getCollation();
             if (collation != null) {
                 sb.append(" collation: ").append(collation);
             }
         }
-        if (sb.length() == 0) {
-            QUERY_LOG.debug("Executing Mongo 'aggregate'");
-        } else {
-            QUERY_LOG.debug("Executing Mongo 'aggregate' with" + sb);
-        }
+        QUERY_LOG.debug(sb.toString());
     }
 
 }

--- a/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultMongoRepositoryOperations.java
+++ b/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultMongoRepositoryOperations.java
@@ -29,7 +29,6 @@ import com.mongodb.client.model.Collation;
 import com.mongodb.client.model.DeleteOneModel;
 import com.mongodb.client.model.Filters;
 import com.mongodb.client.model.ReplaceOneModel;
-import com.mongodb.client.model.ReplaceOptions;
 import com.mongodb.client.model.UpdateOneModel;
 import com.mongodb.client.result.DeleteResult;
 import com.mongodb.client.result.InsertManyResult;
@@ -732,7 +731,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
         if (QUERY_LOG.isDebugEnabled()) {
             QUERY_LOG.debug("Executing Mongo 'insertOne' for collection: {} with document: {}", collection.getNamespace().getFullName(), association);
         }
-        collection.insertOne(ctx.clientSession, association);
+        collection.insertOne(ctx.clientSession, association, getInsertOneOptions(ctx.annotationMetadata));
     }
 
     @Override
@@ -750,7 +749,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
         if (QUERY_LOG.isDebugEnabled()) {
             QUERY_LOG.debug("Executing Mongo 'insertMany' for collection: {} with documents: {}", collection.getNamespace().getFullName(), associations);
         }
-        collection.insertMany(ctx.clientSession, associations);
+        collection.insertMany(ctx.clientSession, associations, getInsertManyOptions(ctx.annotationMetadata));
     }
 
     private <T> T withClientSession(Function<ClientSession, T> function) {
@@ -773,7 +772,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
                 if (QUERY_LOG.isDebugEnabled()) {
                     QUERY_LOG.debug("Executing Mongo 'insertOne' with entity: {}", entity);
                 }
-                InsertOneResult insertOneResult = collection.insertOne(ctx.clientSession, entity);
+                InsertOneResult insertOneResult = collection.insertOne(ctx.clientSession, entity, getInsertOneOptions(ctx.annotationMetadata));
                 BsonValue insertedId = insertOneResult.getInsertedId();
                 BeanProperty<T, Object> property = (BeanProperty<T, Object>) persistentEntity.getIdentity().getProperty();
                 if (property.get(entity) == null) {
@@ -792,7 +791,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
 
             @Override
             protected void collectAutoPopulatedPreviousValues() {
-                filter = MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, entity, collection.getCodecRegistry());
+                filter = createFilterIdAndVersion(persistentEntity, entity, mongoDatabase.getCodecRegistry());
             }
 
             @Override
@@ -802,13 +801,13 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
                 }
                 BsonDocument bsonDocument = BsonDocumentWrapper.asBsonDocument(entity, mongoDatabase.getCodecRegistry());
                 bsonDocument.remove("_id");
-                ReplaceOptions replaceOptions = new ReplaceOptions();
-                UpdateResult updateResult = collection.replaceOne(ctx.clientSession, filter, bsonDocument, replaceOptions);
+                UpdateResult updateResult = collection.replaceOne(ctx.clientSession, filter, bsonDocument, getReplaceOptions(ctx.annotationMetadata));
                 modifiedCount = updateResult.getModifiedCount();
                 if (persistentEntity.getVersion() != null) {
                     checkOptimisticLocking(1, (int) modifiedCount);
                 }
             }
+
         };
     }
 
@@ -854,7 +853,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
             @Override
             protected void collectAutoPopulatedPreviousValues() {
                 filters = entities.stream()
-                        .collect(Collectors.toMap(d -> d, d -> MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, d.entity, collection.getCodecRegistry())));
+                        .collect(Collectors.toMap(d -> d, d -> createFilterIdAndVersion(persistentEntity, d.entity, collection.getCodecRegistry())));
             }
 
             @Override
@@ -870,7 +869,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
                     }
                     BsonDocument bsonDocument = BsonDocumentWrapper.asBsonDocument(d.entity, mongoDatabase.getCodecRegistry());
                     bsonDocument.remove("_id");
-                    replaces.add(new ReplaceOneModel<>(filter, bsonDocument));
+                    replaces.add(new ReplaceOneModel<>(filter, bsonDocument, getReplaceOptions(ctx.annotationMetadata)));
                 }
                 BulkWriteResult bulkWriteResult = collection.bulkWrite(ctx.clientSession, replaces);
                 modifiedCount = bulkWriteResult.getModifiedCount();
@@ -890,7 +889,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
 
             @Override
             protected void collectAutoPopulatedPreviousValues() {
-                filter = MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, entity, collection.getCodecRegistry());
+                filter = createFilterIdAndVersion(persistentEntity, entity, collection.getCodecRegistry());
             }
 
             @Override
@@ -898,7 +897,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
                 if (QUERY_LOG.isDebugEnabled()) {
                     QUERY_LOG.debug("Executing Mongo 'deleteOne' with filter: {}", filter.toBsonDocument().toJson());
                 }
-                DeleteResult deleteResult = collection.deleteOne(ctx.clientSession, filter);
+                DeleteResult deleteResult = collection.deleteOne(ctx.clientSession, filter, getDeleteOptions(ctx.annotationMetadata));
                 modifiedCount = deleteResult.getDeletedCount();
                 if (persistentEntity.getVersion() != null) {
                     checkOptimisticLocking(1, (int) modifiedCount);
@@ -916,7 +915,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
 
             @Override
             protected void collectAutoPopulatedPreviousValues() {
-                filters = entities.stream().collect(Collectors.toMap(d -> d, d -> MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, d.entity, collection.getCodecRegistry())));
+                filters = entities.stream().collect(Collectors.toMap(d -> d, d -> createFilterIdAndVersion(persistentEntity, d.entity, collection.getCodecRegistry())));
             }
 
             @Override
@@ -927,7 +926,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
                     if (QUERY_LOG.isDebugEnabled()) {
                         QUERY_LOG.debug("Executing Mongo 'deleteMany' with filter: {}", filter.toBsonDocument().toJson());
                     }
-                    DeleteResult deleteResult = collection.deleteMany(ctx.clientSession, filter);
+                    DeleteResult deleteResult = collection.deleteMany(ctx.clientSession, filter, getDeleteOptions(ctx.annotationMetadata));
                     modifiedCount = deleteResult.getDeletedCount();
                 }
                 if (persistentEntity.getVersion() != null) {
@@ -980,7 +979,7 @@ public final class DefaultMongoRepositoryOperations extends AbstractMongoReposit
                 }
                 MongoDatabase mongoDatabase = getDatabase(persistentEntity, ctx.repositoryType);
                 InsertManyResult insertManyResult = getCollection(mongoDatabase, persistentEntity, persistentEntity.getIntrospection().getBeanType())
-                        .insertMany(ctx.clientSession, toInsert);
+                        .insertMany(ctx.clientSession, toInsert, getInsertManyOptions(ctx.annotationMetadata));
                 if (hasGeneratedId) {
                     Map<Integer, BsonValue> insertedIds = insertManyResult.getInsertedIds();
                     RuntimePersistentProperty<T> identity = persistentEntity.getIdentity();

--- a/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultMongoStoredQuery.java
+++ b/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultMongoStoredQuery.java
@@ -15,12 +15,10 @@
  */
 package io.micronaut.data.mongodb.operations;
 
-import com.mongodb.CursorType;
 import com.mongodb.client.model.Collation;
 import com.mongodb.client.model.DeleteOptions;
 import com.mongodb.client.model.UpdateOptions;
 import io.micronaut.aop.InvocationContext;
-import io.micronaut.core.annotation.AnnotationValue;
 import io.micronaut.core.annotation.Internal;
 import io.micronaut.core.annotation.NonNull;
 import io.micronaut.core.annotation.Nullable;
@@ -50,6 +48,7 @@ import io.micronaut.data.mongodb.annotation.MongoProjection;
 import io.micronaut.data.mongodb.annotation.MongoSort;
 import io.micronaut.data.mongodb.operations.options.MongoAggregationOptions;
 import io.micronaut.data.mongodb.operations.options.MongoFindOptions;
+import io.micronaut.data.mongodb.operations.options.MongoOptionsUtils;
 import io.micronaut.data.runtime.query.internal.DefaultStoredQuery;
 import io.micronaut.data.runtime.query.internal.DelegateStoredQuery;
 import org.bson.BsonArray;
@@ -585,19 +584,7 @@ final class DefaultMongoStoredQuery<E, R, Dtb> implements DelegateStoredQuery<E,
             this.pipelineParameterIndex = getParameterIndexByName(pipelineParameter);
             this.optionsParameterIndex = getParameterIndexByName(optionsParameter);
             this.pipelineNeedsProcessing = needsProcessing(pipeline);
-            AnnotationValue<io.micronaut.data.mongodb.annotation.MongoAggregateOptions> optionsAnn = storedQuery.getAnnotationMetadata().getAnnotation(io.micronaut.data.mongodb.annotation.MongoAggregateOptions.class);
-            if (optionsAnn != null) {
-                MongoAggregationOptions options = new MongoAggregationOptions();
-                optionsAnn.booleanValue("bypassDocumentValidation").ifPresent(options::bypassDocumentValidation);
-                optionsAnn.longValue("maxTimeMS").ifPresent(options::maxTimeMS);
-                optionsAnn.longValue("maxAwaitTimeMS").ifPresent(options::maxAwaitTimeMS);
-                optionsAnn.stringValue("comment").ifPresent(options::comment);
-                optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
-                optionsAnn.booleanValue("allowDiskUse").ifPresent(options::allowDiskUse);
-                this.options = options.isEmpty() ? null : options;
-            } else {
-                options = null;
-            }
+            options = MongoOptionsUtils.buildAggregateOptions(storedQuery.getAnnotationMetadata()).orElse(null);
         }
 
         public MongoAggregation getAggregation(InvocationContext<?, ?> invocationContext) {
@@ -656,15 +643,7 @@ final class DefaultMongoStoredQuery<E, R, Dtb> implements DelegateStoredQuery<E,
             this.filterParameterIndex = getParameterIndexByName(filterParameter);
             this.updateParameterIndex = getParameterIndexByName(updateParameter);
             this.optionsParameterIndex = getParameterIndexByName(optionsParameter);
-            AnnotationValue<io.micronaut.data.mongodb.annotation.MongoUpdateOptions> optionsAnn = storedQuery.getAnnotationMetadata().getAnnotation(io.micronaut.data.mongodb.annotation.MongoUpdateOptions.class);
-            if (optionsAnn != null) {
-                options = new UpdateOptions();
-                optionsAnn.booleanValue("upsert").ifPresent(options::upsert);
-                optionsAnn.booleanValue("bypassDocumentValidation").ifPresent(options::bypassDocumentValidation);
-                optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
-            } else {
-                options = null;
-            }
+            this.options = MongoOptionsUtils.buildUpdateOptions(storedQuery.getAnnotationMetadata(), false).orElse(null);
         }
 
         private UpdateOptions copy(UpdateOptions options) {
@@ -790,29 +769,7 @@ final class DefaultMongoStoredQuery<E, R, Dtb> implements DelegateStoredQuery<E,
             projectionNeedsProcessing = needsProcessing(projection);
             this.filter = filter;
             this.filterNeedsProcessing = needsProcessing(filter);
-
-            AnnotationValue<io.micronaut.data.mongodb.annotation.MongoFindOptions> optionsAnn = storedQuery.getAnnotationMetadata().getAnnotation(io.micronaut.data.mongodb.annotation.MongoFindOptions.class);
-            if (optionsAnn != null) {
-                MongoFindOptions options = new MongoFindOptions();
-                optionsAnn.intValue("batchSize").ifPresent(options::batchSize);
-                optionsAnn.intValue("skip").ifPresent(options::skip);
-                optionsAnn.intValue("limit").ifPresent(options::limit);
-                optionsAnn.longValue("maxTimeMS").ifPresent(options::maxTimeMS);
-                optionsAnn.longValue("maxAwaitTimeMS").ifPresent(options::maxAwaitTimeMS);
-                optionsAnn.enumValue("cursorType", CursorType.class).ifPresent(options::cursorType);
-                optionsAnn.booleanValue("noCursorTimeout").ifPresent(options::noCursorTimeout);
-                optionsAnn.booleanValue("partial").ifPresent(options::partial);
-                optionsAnn.stringValue("comment").ifPresent(options::comment);
-                optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
-                optionsAnn.stringValue("max").map(BsonDocument::parse).ifPresent(options::max);
-                optionsAnn.stringValue("min").map(BsonDocument::parse).ifPresent(options::min);
-                optionsAnn.booleanValue("returnKey").ifPresent(options::returnKey);
-                optionsAnn.booleanValue("showRecordId").ifPresent(options::showRecordId);
-                optionsAnn.booleanValue("allowDiskUse").ifPresent(options::allowDiskUse);
-                this.options = options.isEmpty() ? null : options;
-            } else {
-                options = null;
-            }
+            options = MongoOptionsUtils.buildFindOptions(storedQuery.getAnnotationMetadata()).orElse(null);
         }
 
         public MongoFind getFind(InvocationContext<?, ?> invocationContext) {
@@ -894,13 +851,7 @@ final class DefaultMongoStoredQuery<E, R, Dtb> implements DelegateStoredQuery<E,
             this.filterNeedsProcessing = needsProcessing(filter);
             this.filterParameterIndex = getParameterIndexByName(filterParameter);
             this.optionsParameterIndex = getParameterIndexByName(optionsParameter);
-            AnnotationValue<io.micronaut.data.mongodb.annotation.MongoDeleteOptions> optionsAnn = storedQuery.getAnnotationMetadata().getAnnotation(io.micronaut.data.mongodb.annotation.MongoDeleteOptions.class);
-            if (optionsAnn != null) {
-                options = new DeleteOptions();
-                optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
-            } else {
-                options = null;
-            }
+            options = MongoOptionsUtils.buildDeleteOptions(storedQuery.getAnnotationMetadata(), false).orElse(null);
         }
 
         public MongoDelete getDeleteMany(InvocationContext<?, ?> invocationContext) {
@@ -976,7 +927,7 @@ final class DefaultMongoStoredQuery<E, R, Dtb> implements DelegateStoredQuery<E,
         protected CollationSupported() {
             collationAsBson = storedQuery.getAnnotationMetadata().stringValue(MongoCollation.class).map(BsonDocument::parse).orElse(null);
             collationNeedsProcessing = needsProcessing(collationAsBson);
-            collation = collationAsBson == null || collationNeedsProcessing ? null : MongoUtils.bsonDocumentAsCollation(collationAsBson.toBsonDocument());
+            collation = collationAsBson == null || collationNeedsProcessing ? null : MongoOptionsUtils.bsonDocumentAsCollation(collationAsBson.toBsonDocument());
         }
 
         protected Collation getCollation(@Nullable InvocationContext<?, ?> invocationContext, @Nullable E entity) {
@@ -987,7 +938,7 @@ final class DefaultMongoStoredQuery<E, R, Dtb> implements DelegateStoredQuery<E,
                 return null;
             }
             Bson collationAsBson = collationNeedsProcessing ? replaceQueryParameters(this.collationAsBson, invocationContext, entity) : this.collationAsBson;
-            return MongoUtils.bsonDocumentAsCollation(collationAsBson.toBsonDocument());
+            return MongoOptionsUtils.bsonDocumentAsCollation(collationAsBson.toBsonDocument());
         }
     }
 }

--- a/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultReactiveMongoRepositoryOperations.java
+++ b/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/DefaultReactiveMongoRepositoryOperations.java
@@ -681,7 +681,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
         if (QUERY_LOG.isDebugEnabled()) {
             QUERY_LOG.debug("Executing Mongo 'insertOne' for collection: {} with document: {}", collection.getNamespace().getFullName(), association);
         }
-        return Mono.from(collection.insertOne(ctx.clientSession, association)).then();
+        return Mono.from(collection.insertOne(ctx.clientSession, association, getInsertOneOptions(ctx.annotationMetadata))).then();
     }
 
     @Override
@@ -695,7 +695,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
         if (QUERY_LOG.isDebugEnabled()) {
             QUERY_LOG.debug("Executing Mongo 'insertMany' for collection: {} with associations: {}", collection.getNamespace().getFullName(), associations);
         }
-        return Mono.from(collection.insertMany(ctx.clientSession, associations)).then();
+        return Mono.from(collection.insertMany(ctx.clientSession, associations, getInsertManyOptions(ctx.annotationMetadata))).then();
     }
 
     @Override
@@ -755,7 +755,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     if (QUERY_LOG.isDebugEnabled()) {
                         QUERY_LOG.debug("Executing Mongo 'insertOne' with entity: {}", d.entity);
                     }
-                    return Mono.from(collection.insertOne(ctx.clientSession, d.entity)).map(insertOneResult -> {
+                    return Mono.from(collection.insertOne(ctx.clientSession, d.entity, getInsertOneOptions(ctx.annotationMetadata))).map(insertOneResult -> {
                         BsonValue insertedId = insertOneResult.getInsertedId();
                         BeanProperty<T, Object> property = (BeanProperty<T, Object>) persistentEntity.getIdentity().getProperty();
                         if (property.get(d.entity) == null) {
@@ -777,7 +777,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
             @Override
             protected void collectAutoPopulatedPreviousValues() {
                 data = data.map(d -> {
-                    d.filter = MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, d.entity, collection.getCodecRegistry());
+                    d.filter = createFilterIdAndVersion(persistentEntity, d.entity, collection.getCodecRegistry());
                     return d;
                 });
             }
@@ -791,7 +791,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     }
                     BsonDocument bsonDocument = BsonDocumentWrapper.asBsonDocument(d.entity, mongoDatabase.getCodecRegistry());
                     bsonDocument.remove("_id");
-                    return Mono.from(collection.replaceOne(ctx.clientSession, filter, bsonDocument)).map(updateResult -> {
+                    return Mono.from(collection.replaceOne(ctx.clientSession, filter, bsonDocument, getReplaceOptions(ctx.annotationMetadata))).map(updateResult -> {
                         d.rowsUpdated = updateResult.getModifiedCount();
                         if (persistentEntity.getVersion() != null) {
                             checkOptimisticLocking(1, (int) d.rowsUpdated);
@@ -817,7 +817,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     if (d.vetoed) {
                         return d;
                     }
-                    d.filter = MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, d.entity, collection.getCodecRegistry());
+                    d.filter = createFilterIdAndVersion(persistentEntity, d.entity, collection.getCodecRegistry());
                     return d;
                 });
             }
@@ -836,7 +836,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                         }
                         BsonDocument bsonDocument = BsonDocumentWrapper.asBsonDocument(d.entity, mongoDatabase.getCodecRegistry());
                         bsonDocument.remove("_id");
-                        replaces.add(new ReplaceOneModel<>(filter, bsonDocument));
+                        replaces.add(new ReplaceOneModel<>(filter, bsonDocument, getReplaceOptions(ctx.annotationMetadata)));
                     }
                     return Mono.from(collection.bulkWrite(ctx.clientSession, replaces)).map(bulkWriteResult -> {
                         if (persistentEntity.getVersion() != null) {
@@ -894,7 +894,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     if (d.vetoed) {
                         return d;
                     }
-                    d.filter = MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, d.entity, collection.getCodecRegistry());
+                    d.filter = createFilterIdAndVersion(persistentEntity, d.entity, collection.getCodecRegistry());
                     return d;
                 });
             }
@@ -906,7 +906,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     if (QUERY_LOG.isDebugEnabled()) {
                         QUERY_LOG.debug("Executing Mongo 'deleteOne' with filter: {}", filter.toBsonDocument().toJson());
                     }
-                    return Mono.from(getCollection(persistentEntity, ctx.repositoryType).deleteOne(ctx.clientSession, filter)).map(deleteResult -> {
+                    return Mono.from(getCollection(persistentEntity, ctx.repositoryType).deleteOne(ctx.clientSession, filter, getDeleteOptions(ctx.annotationMetadata))).map(deleteResult -> {
                         d.rowsUpdated = (int) deleteResult.getDeletedCount();
                         if (persistentEntity.getVersion() != null) {
                             checkOptimisticLocking(1, d.rowsUpdated);
@@ -932,7 +932,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     if (d.vetoed) {
                         return d;
                     }
-                    d.filter = MongoUtils.filterByIdAndVersion(conversionService, persistentEntity, d.entity, collection.getCodecRegistry());
+                    d.filter = createFilterIdAndVersion(persistentEntity, d.entity, collection.getCodecRegistry());
                     return d;
                 });
             }
@@ -947,7 +947,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                         if (QUERY_LOG.isDebugEnabled()) {
                             QUERY_LOG.debug("Executing Mongo 'deleteMany' with filter: {}", filter.toBsonDocument().toJson());
                         }
-                        modifiedCount = Mono.from(collection.deleteMany(ctx.clientSession, filter)).map(DeleteResult::getDeletedCount);
+                        modifiedCount = Mono.from(collection.deleteMany(ctx.clientSession, filter, getDeleteOptions(ctx.annotationMetadata))).map(DeleteResult::getDeletedCount);
                     } else {
                         modifiedCount = Mono.just(0L);
                     }
@@ -1007,7 +1007,7 @@ public class DefaultReactiveMongoRepositoryOperations extends AbstractMongoRepos
                     if (QUERY_LOG.isDebugEnabled()) {
                         QUERY_LOG.debug("Executing Mongo 'insertMany' for collection: {} with documents: {}", collection.getNamespace().getFullName(), toInsert);
                     }
-                    return Mono.from(collection.insertMany(ctx.clientSession, toInsert)).flatMapMany(insertManyResult -> {
+                    return Mono.from(collection.insertMany(ctx.clientSession, toInsert, getInsertManyOptions(ctx.annotationMetadata))).flatMapMany(insertManyResult -> {
                         if (hasGeneratedId) {
                             Map<Integer, BsonValue> insertedIds = insertManyResult.getInsertedIds();
                             RuntimePersistentProperty<T> identity = persistentEntity.getIdentity();

--- a/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/MongoUtils.java
+++ b/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/MongoUtils.java
@@ -15,22 +15,13 @@
  */
 package io.micronaut.data.mongodb.operations;
 
-import com.mongodb.client.model.Collation;
-import com.mongodb.client.model.CollationAlternate;
-import com.mongodb.client.model.CollationCaseFirst;
-import com.mongodb.client.model.CollationMaxVariable;
-import com.mongodb.client.model.CollationStrength;
-import com.mongodb.client.model.Filters;
 import io.micronaut.core.annotation.AnnotationValue;
 import io.micronaut.core.annotation.Internal;
-import io.micronaut.core.annotation.Nullable;
 import io.micronaut.core.beans.BeanProperty;
 import io.micronaut.core.convert.ConversionService;
 import io.micronaut.data.model.Association;
-import io.micronaut.data.model.PersistentProperty;
 import io.micronaut.data.model.runtime.RuntimePersistentEntity;
 import io.micronaut.data.model.runtime.RuntimePersistentProperty;
-import io.micronaut.serde.config.annotation.SerdeConfig;
 import org.bson.BsonDocument;
 import org.bson.BsonDocumentWrapper;
 import org.bson.BsonInt32;
@@ -56,48 +47,7 @@ public final class MongoUtils {
 
     public static final String ID = "_id";
 
-    public static Collation bsonDocumentAsCollation(@Nullable BsonDocument collationDocument) {
-        if (collationDocument == null) {
-            return null;
-        }
-        Collation.Builder builder = Collation.builder();
-        BsonValue locale = collationDocument.get("locale");
-        if (locale != null) {
-            builder.locale(locale.asString().getValue());
-        }
-        BsonValue caseLevel = collationDocument.get("caseLevel");
-        if (caseLevel != null) {
-            builder.caseLevel(caseLevel.asBoolean().getValue());
-        }
-        BsonValue caseFirst = collationDocument.get("caseFirst");
-        if (caseFirst != null) {
-            builder.collationCaseFirst(CollationCaseFirst.valueOf(caseFirst.asString().getValue()));
-        }
-        BsonValue strength = collationDocument.get("strength");
-        if (strength != null) {
-            builder.collationStrength(CollationStrength.valueOf(strength.asString().getValue()));
-        }
-        BsonValue numericOrdering = collationDocument.get("numericOrdering");
-        if (numericOrdering != null) {
-            builder.numericOrdering(numericOrdering.asBoolean().getValue());
-        }
-        BsonValue alternate = collationDocument.get("alternate");
-        if (alternate != null) {
-            builder.collationAlternate(CollationAlternate.valueOf(alternate.asString().getValue()));
-        }
-        BsonValue maxVariable = collationDocument.get("maxVariable");
-        if (maxVariable != null) {
-            builder.collationMaxVariable(CollationMaxVariable.valueOf(maxVariable.asString().getValue()));
-        }
-        BsonValue normalization = collationDocument.get("normalization");
-        if (normalization != null) {
-            builder.normalization(normalization.asBoolean().getValue());
-        }
-        BsonValue backwards = collationDocument.get("backwards");
-        if (backwards != null) {
-            builder.backwards(backwards.asBoolean().getValue());
-        }
-        return builder.build();
+    private MongoUtils() {
     }
 
     public static BsonValue entityIdValue(ConversionService<?> conversionService,
@@ -143,34 +93,6 @@ public final class MongoUtils {
                            CodecRegistry codecRegistry) {
         BsonValue id = idValue(conversionService, persistentEntity, value, codecRegistry);
         return new BsonDocument().append(ID, id);
-    }
-
-    static Bson filterByEntityId(ConversionService<?> conversionService,
-                                 RuntimePersistentEntity<?> persistentEntity,
-                                 Object entity,
-                                 CodecRegistry codecRegistry) {
-        BsonValue id = entityIdValue(conversionService, persistentEntity, entity, codecRegistry);
-        return new BsonDocument().append(ID, id);
-    }
-
-    static Bson filterByIdAndVersion(ConversionService<?> conversionService,
-                                     RuntimePersistentEntity persistentEntity,
-                                     Object entity,
-                                     CodecRegistry codecRegistry) {
-        RuntimePersistentProperty version = persistentEntity.getVersion();
-        if (version != null) {
-            return Filters.and(
-                    filterByEntityId(conversionService, persistentEntity, entity, codecRegistry),
-                    Filters.eq(getPropertyPersistName(version), version.getProperty().get(entity))
-            );
-        }
-        return filterByEntityId(conversionService, persistentEntity, entity, codecRegistry);
-    }
-
-    private static String getPropertyPersistName(PersistentProperty property) {
-        return property.getAnnotationMetadata()
-                .stringValue(SerdeConfig.class, SerdeConfig.PROPERTY)
-                .orElseGet(property::getName);
     }
 
     static <T> T toValue(BsonDocument bsonDocument, Class<T> resultClass, CodecRegistry codecRegistry) {

--- a/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/options/MongoOptionsUtils.java
+++ b/data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/options/MongoOptionsUtils.java
@@ -1,0 +1,198 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.mongodb.operations.options;
+
+import com.mongodb.CursorType;
+import com.mongodb.client.model.Collation;
+import com.mongodb.client.model.CollationAlternate;
+import com.mongodb.client.model.CollationCaseFirst;
+import com.mongodb.client.model.CollationMaxVariable;
+import com.mongodb.client.model.CollationStrength;
+import com.mongodb.client.model.DeleteOptions;
+import com.mongodb.client.model.InsertManyOptions;
+import com.mongodb.client.model.InsertOneOptions;
+import com.mongodb.client.model.ReplaceOptions;
+import com.mongodb.client.model.UpdateOptions;
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.data.mongodb.annotation.MongoCollation;
+import io.micronaut.data.mongodb.annotation.MongoDeleteOptions;
+import io.micronaut.data.mongodb.annotation.MongoUpdateOptions;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+
+import java.util.Optional;
+
+/**
+ * Mongo internal utils for building options.
+ */
+@Internal
+public final class MongoOptionsUtils {
+
+    private MongoOptionsUtils() {
+    }
+
+    public static Optional<UpdateOptions> buildUpdateOptions(AnnotationMetadata annotationMetadata, boolean includeCollation) {
+        AnnotationValue<MongoUpdateOptions> optionsAnn = annotationMetadata.getAnnotation(MongoUpdateOptions.class);
+        if (optionsAnn == null) {
+            return Optional.empty();
+        }
+        UpdateOptions options = new UpdateOptions();
+        optionsAnn.booleanValue("upsert").ifPresent(options::upsert);
+        optionsAnn.booleanValue("bypassDocumentValidation").ifPresent(options::bypassDocumentValidation);
+        optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
+        if (includeCollation) {
+            annotationMetadata.stringValue(MongoCollation.class)
+                    .map(BsonDocument::parse)
+                    .ifPresent(bsonDocument -> options.collation(bsonDocumentAsCollation(bsonDocument)));
+
+        }
+        return Optional.of(options);
+    }
+
+    public static Optional<ReplaceOptions> buildReplaceOptions(AnnotationMetadata annotationMetadata) {
+        AnnotationValue<MongoUpdateOptions> optionsAnn = annotationMetadata.getAnnotation(MongoUpdateOptions.class);
+        if (optionsAnn == null) {
+            return Optional.empty();
+        }
+        ReplaceOptions options = new ReplaceOptions();
+        optionsAnn.booleanValue("upsert").ifPresent(options::upsert);
+        optionsAnn.booleanValue("bypassDocumentValidation").ifPresent(options::bypassDocumentValidation);
+        optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
+        annotationMetadata.stringValue(MongoCollation.class)
+                .map(BsonDocument::parse)
+                .ifPresent(bsonDocument -> options.collation(bsonDocumentAsCollation(bsonDocument)));
+        return Optional.of(options);
+    }
+
+    public static Optional<InsertOneOptions> buildInsertOneOptions(AnnotationMetadata annotationMetadata) {
+        // Future annotation support
+        return Optional.empty();
+    }
+
+    public static Optional<InsertManyOptions> buildInsertManyOptions(AnnotationMetadata annotationMetadata) {
+        // Future annotation support
+        return Optional.empty();
+    }
+
+    public static Optional<MongoFindOptions> buildFindOptions(AnnotationMetadata annotationMetadata) {
+        AnnotationValue<io.micronaut.data.mongodb.annotation.MongoFindOptions> optionsAnn = annotationMetadata
+                .getAnnotation(io.micronaut.data.mongodb.annotation.MongoFindOptions.class);
+        if (optionsAnn == null) {
+            return Optional.empty();
+        }
+        MongoFindOptions options = new MongoFindOptions();
+        optionsAnn.intValue("batchSize").ifPresent(options::batchSize);
+        optionsAnn.intValue("skip").ifPresent(options::skip);
+        optionsAnn.intValue("limit").ifPresent(options::limit);
+        optionsAnn.longValue("maxTimeMS").ifPresent(options::maxTimeMS);
+        optionsAnn.longValue("maxAwaitTimeMS").ifPresent(options::maxAwaitTimeMS);
+        optionsAnn.enumValue("cursorType", CursorType.class).ifPresent(options::cursorType);
+        optionsAnn.booleanValue("noCursorTimeout").ifPresent(options::noCursorTimeout);
+        optionsAnn.booleanValue("partial").ifPresent(options::partial);
+        optionsAnn.stringValue("comment").ifPresent(options::comment);
+        optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
+        optionsAnn.stringValue("max").map(BsonDocument::parse).ifPresent(options::max);
+        optionsAnn.stringValue("min").map(BsonDocument::parse).ifPresent(options::min);
+        optionsAnn.booleanValue("returnKey").ifPresent(options::returnKey);
+        optionsAnn.booleanValue("showRecordId").ifPresent(options::showRecordId);
+        optionsAnn.booleanValue("allowDiskUse").ifPresent(options::allowDiskUse);
+        if (options.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(options);
+    }
+
+    public static Optional<MongoAggregationOptions> buildAggregateOptions(AnnotationMetadata annotationMetadata) {
+        AnnotationValue<io.micronaut.data.mongodb.annotation.MongoAggregateOptions> optionsAnn = annotationMetadata
+                .getAnnotation(io.micronaut.data.mongodb.annotation.MongoAggregateOptions.class);
+        if (optionsAnn == null) {
+            return Optional.empty();
+        }
+        MongoAggregationOptions options = new MongoAggregationOptions();
+        optionsAnn.booleanValue("bypassDocumentValidation").ifPresent(options::bypassDocumentValidation);
+        optionsAnn.longValue("maxTimeMS").ifPresent(options::maxTimeMS);
+        optionsAnn.longValue("maxAwaitTimeMS").ifPresent(options::maxAwaitTimeMS);
+        optionsAnn.stringValue("comment").ifPresent(options::comment);
+        optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
+        optionsAnn.booleanValue("allowDiskUse").ifPresent(options::allowDiskUse);
+        if (options.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(options);
+    }
+
+    public static Optional<DeleteOptions> buildDeleteOptions(AnnotationMetadata annotationMetadata, boolean includeCollation) {
+        AnnotationValue<MongoDeleteOptions> optionsAnn = annotationMetadata
+                .getAnnotation(MongoDeleteOptions.class);
+        if (optionsAnn == null) {
+            return Optional.empty();
+        }
+        DeleteOptions options = new DeleteOptions();
+        optionsAnn.stringValue("hint").map(BsonDocument::parse).ifPresent(options::hint);
+        annotationMetadata.stringValue(MongoCollation.class)
+                .map(BsonDocument::parse)
+                .ifPresent(bsonDocument -> options.collation(bsonDocumentAsCollation(bsonDocument)));
+        return Optional.of(options);
+    }
+
+    public static Collation bsonDocumentAsCollation(@Nullable BsonDocument collationDocument) {
+        if (collationDocument == null) {
+            return null;
+        }
+        Collation.Builder builder = Collation.builder();
+        BsonValue locale = collationDocument.get("locale");
+        if (locale != null) {
+            builder.locale(locale.asString().getValue());
+        }
+        BsonValue caseLevel = collationDocument.get("caseLevel");
+        if (caseLevel != null) {
+            builder.caseLevel(caseLevel.asBoolean().getValue());
+        }
+        BsonValue caseFirst = collationDocument.get("caseFirst");
+        if (caseFirst != null) {
+            builder.collationCaseFirst(CollationCaseFirst.valueOf(caseFirst.asString().getValue()));
+        }
+        BsonValue strength = collationDocument.get("strength");
+        if (strength != null) {
+            builder.collationStrength(CollationStrength.valueOf(strength.asString().getValue()));
+        }
+        BsonValue numericOrdering = collationDocument.get("numericOrdering");
+        if (numericOrdering != null) {
+            builder.numericOrdering(numericOrdering.asBoolean().getValue());
+        }
+        BsonValue alternate = collationDocument.get("alternate");
+        if (alternate != null) {
+            builder.collationAlternate(CollationAlternate.valueOf(alternate.asString().getValue()));
+        }
+        BsonValue maxVariable = collationDocument.get("maxVariable");
+        if (maxVariable != null) {
+            builder.collationMaxVariable(CollationMaxVariable.valueOf(maxVariable.asString().getValue()));
+        }
+        BsonValue normalization = collationDocument.get("normalization");
+        if (normalization != null) {
+            builder.normalization(normalization.asBoolean().getValue());
+        }
+        BsonValue backwards = collationDocument.get("backwards");
+        if (backwards != null) {
+            builder.backwards(backwards.asBoolean().getValue());
+        }
+        return builder.build();
+    }
+
+}

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/MongoIdsSpec.groovy
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/MongoIdsSpec.groovy
@@ -58,7 +58,7 @@ class MongoIdsSpec extends Specification implements MongoTestPropertyProvider {
             def one = dbRepositoryOperations.findOne(EntityObjectId, persisted.myId)
         then:
             one.myId == persisted.myId
-            when:
+        when:
             def found = database.getCollection("entities_object_ids", BsonDocument).find().first()
         then:
             found._id instanceof BsonObjectId

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/SimpleSongSpec.groovy
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/SimpleSongSpec.groovy
@@ -1,0 +1,62 @@
+package io.micronaut.data.document.mongodb.upsert
+
+import io.micronaut.data.document.mongodb.MongoTestPropertyProvider
+import io.micronaut.data.document.mongodb.upsert.model.SongEntity
+import io.micronaut.data.document.mongodb.upsert.repo.SongRepository
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import jakarta.inject.Inject
+import org.junit.Test
+import spock.lang.Specification
+
+import java.time.LocalDateTime
+import java.time.temporal.ChronoUnit
+
+/**
+ * The overall goal is to use Strings as keys and:
+ * to write to the repository without having to read the entity first
+ * (insert if not exist, always overwrite if same id exists)
+ */
+@MicronautTest(transactional = false)
+class SimpleSongSpec extends Specification implements MongoTestPropertyProvider {
+
+    @Inject
+    private SongRepository songRepository
+
+    @Test
+    void savesAndRetrievesSongs() {
+        when:
+            SongEntity songEntity = new SongEntity()
+            songEntity.setSongHash("song_1")
+            songEntity.setName("Don't Worry, be happy")
+            songRepository.save(songEntity)
+            LocalDateTime created = songEntity.getCreated().truncatedTo(ChronoUnit.MILLIS)
+            LocalDateTime updated = songEntity.getUpdated().truncatedTo(ChronoUnit.MILLIS)
+        then:
+            SongEntity song1 = songRepository.findById("song_1").get()
+            song1.getSongHash() == "song_1"
+
+        when:
+            def updatedSong = songRepository.update(song1)
+            def updatedSong2 = songRepository.findById("song_1").get()
+        then:
+            updatedSong.created == created
+            updatedSong.updated != updated
+            updatedSong.updated == song1.updated
+            updatedSong2.created == created
+            updatedSong2.updated != updated
+            updatedSong2.updated == updatedSong.updated.truncatedTo(ChronoUnit.MILLIS)
+    }
+
+    @Test
+    void updatesSongs() {
+        when:
+            SongEntity songEntity = new SongEntity()
+            songEntity.setSongHash("song_2")
+            songEntity.setName("Don't Worry, be happy")
+
+            songRepository.update(songEntity)
+        then:
+            Optional<SongEntity> song2 = songRepository.findById("song_2")
+            song2.isPresent()
+    }
+}

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/SimpleSongSpec2.groovy
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/SimpleSongSpec2.groovy
@@ -1,0 +1,56 @@
+package io.micronaut.data.document.mongodb.upsert
+
+import io.micronaut.data.document.mongodb.MongoTestPropertyProvider
+import io.micronaut.data.document.mongodb.upsert.model.SongEntity2
+import io.micronaut.data.document.mongodb.upsert.repo.SongRepository2
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import jakarta.inject.Inject
+import org.bson.types.ObjectId
+import org.junit.Test
+import spock.lang.Specification
+
+/**
+ * The overall goal is to use Strings as keys and:
+ * to write to the repository without having to read the entity first
+ * (insert if not exist, always overwrite if same id exists)
+ *
+ * This version uses ObjectIds (which we would rather not use because our
+ * ids have 32 bytes while ObjectIds only allow 24)
+ *
+ */
+@MicronautTest(transactional = false)
+class SimpleSongSpec2 extends Specification implements MongoTestPropertyProvider {
+
+    @Inject
+    private SongRepository2 songRepository
+
+    @Test
+    void savesAndRetrievesSongs() {
+        when:
+            ObjectId id = new ObjectId("ffffffffffffffffffffffff")
+            SongEntity2 songEntity = new SongEntity2()
+            songEntity.setId(id)
+            songEntity.setName("Don't Worry, be happy")
+
+            songRepository.save(songEntity)
+        then:
+            Optional<SongEntity2> song1 = songRepository.findById(id)
+            song1.isPresent()
+    }
+
+    @Test
+    void updatesSongs() {
+        when:
+            ObjectId id = new ObjectId("eeeeeeeeeeeeeeeeeeeeeeee")
+
+            SongEntity2 songEntity = new SongEntity2()
+            songEntity.setId(id)
+            songEntity.setName("Don't Worry, be happy")
+
+            songRepository.update(songEntity)
+        then:
+            Optional<SongEntity2> song2 = songRepository.findById(id)
+            song2.isPresent()
+    }
+
+}

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/model/SongEntity.java
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/model/SongEntity.java
@@ -1,0 +1,53 @@
+package io.micronaut.data.document.mongodb.upsert.model;
+
+import io.micronaut.data.annotation.DateCreated;
+import io.micronaut.data.annotation.DateUpdated;
+import io.micronaut.data.annotation.Id;
+import io.micronaut.data.annotation.MappedEntity;
+
+import java.time.LocalDateTime;
+
+@MappedEntity("songs")
+public class SongEntity {
+
+    @Id
+    String songHash;
+    String name;
+
+    @DateCreated
+    LocalDateTime created;
+    @DateUpdated
+    LocalDateTime updated;
+
+    public String getSongHash() {
+        return songHash;
+    }
+
+    public void setSongHash(String songHash) {
+        this.songHash = songHash;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public LocalDateTime getCreated() {
+        return created;
+    }
+
+    public void setCreated(LocalDateTime created) {
+        this.created = created;
+    }
+
+    public LocalDateTime getUpdated() {
+        return updated;
+    }
+
+    public void setUpdated(LocalDateTime updated) {
+        this.updated = updated;
+    }
+}

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/model/SongEntity2.java
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/model/SongEntity2.java
@@ -1,0 +1,29 @@
+package io.micronaut.data.document.mongodb.upsert.model;
+
+import io.micronaut.data.annotation.Id;
+import io.micronaut.data.annotation.MappedEntity;
+import org.bson.types.ObjectId;
+
+@MappedEntity("songs2")
+public class SongEntity2 {
+
+    @Id
+    ObjectId id;
+    String name;
+
+    public ObjectId getId() {
+        return id;
+    }
+
+    public void setId(ObjectId id) {
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+}

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/repo/SongRepository.java
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/repo/SongRepository.java
@@ -1,0 +1,19 @@
+package io.micronaut.data.document.mongodb.upsert.repo;
+
+import io.micronaut.data.document.mongodb.upsert.model.SongEntity;
+import io.micronaut.data.mongodb.annotation.MongoRepository;
+import io.micronaut.data.mongodb.annotation.MongoUpdateOptions;
+import io.micronaut.data.repository.CrudRepository;
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+@MongoRepository
+public abstract class SongRepository implements CrudRepository<SongEntity, String> {
+    @Override
+    @MongoUpdateOptions(upsert = true)
+    public abstract <S extends SongEntity> S update(@Valid @NotNull S entity);
+
+    @MongoUpdateOptions(upsert = true)
+    public abstract <S extends SongEntity> S save(@Valid @NotNull S entity);
+}

--- a/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/repo/SongRepository2.java
+++ b/data-mongodb/src/test/groovy/io/micronaut/data/document/mongodb/upsert/repo/SongRepository2.java
@@ -1,0 +1,20 @@
+package io.micronaut.data.document.mongodb.upsert.repo;
+
+import io.micronaut.data.document.mongodb.upsert.model.SongEntity2;
+import io.micronaut.data.mongodb.annotation.MongoRepository;
+import io.micronaut.data.mongodb.annotation.MongoUpdateOptions;
+import io.micronaut.data.repository.CrudRepository;
+import org.bson.types.ObjectId;
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+@MongoRepository
+public abstract class SongRepository2 implements CrudRepository<SongEntity2, ObjectId> {
+    @Override
+    @MongoUpdateOptions(upsert = true)
+    public abstract <S extends SongEntity2> S update(@Valid @NotNull S entity);
+
+    @MongoUpdateOptions(upsert = true)
+    public abstract <S extends SongEntity2> S save(@Valid @NotNull S entity);
+}

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/event/listeners/AutoTimestampEntityEventListener.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/event/listeners/AutoTimestampEntityEventListener.java
@@ -95,7 +95,7 @@ public class AutoTimestampEntityEventListener extends AutoPopulatedEntityEventLi
         final Object now = dateTimeProvider.getNow();
         for (RuntimePersistentProperty<Object> property : applicableProperties) {
             if (isUpdate) {
-                if (!property.getAnnotationMetadata().booleanValue(AutoPopulated.class, "updateable").orElse(true)) {
+                if (!property.getAnnotationMetadata().booleanValue(AutoPopulated.class, AutoPopulated.UPDATEABLE).orElse(true)) {
                     continue;
                 }
             }

